Sources/AsyncHTTPClient/ConnectionPool/HTTPExecutingRequest.swift

```diff
@@ -0,0 +1,249 @@
+    func fail(_ error: Error)
+}
+
+/// A handle to the request executor.
+///
+/// This protocol is implemented by the `HTTP1ClientChannelHandler`.
+protocol HTTPRequestExecutor {
+    /// Writes a body part into the channel pipeline
+    ///
+    /// This method may be **called on any thread**. The executor needs to ensure thread safety.
+    func writeRequestBodyPart(_: IOData, request: HTTPExecutingRequest)
+
+    /// Signals that the request body stream has finished
+    ///
+    /// This method may be **called on any thread**. The executor needs to ensure thread safety.
+    func finishRequestBodyStream(_ task: HTTPExecutingRequest)
+
+    /// Signals that more bytes from response body stream can be consumed.
+    ///
+    /// The request executor will call `receiveResponseBodyPart(_ buffer: ByteBuffer)` with more data after
+    /// this call.
+    ///
+    /// This method may be **called on any thread**. The executor needs to ensure thread safety.
+    func demandResponseBodyStream(_ task: HTTPExecutingRequest)
+
+    /// Signals that the request has been cancelled.
+    ///
+    /// This method may be **called on any thread**. The executor needs to ensure thread safety.
+    func cancelRequest(_ task: HTTPExecutingRequest)
+}
+
+protocol HTTPExecutingRequest: AnyObject {
+    /// The request's head.
+    ///
+    /// Based on the content of the request head the task executor will call `startRequestBodyStream`
+    /// after `requestHeadSent` was called.
+    var requestHead: HTTPRequestHead { get }
+
+    /// The maximal `TimeAmount` that is allowed to pass between reads from the Channel.
+    var idleReadTimeout: TimeAmount? { get }
+
+    /// Will be called by the ChannelHandler to indicate that the request is going to be send.
+    ///
+    /// This will be called on the Channel's EventLoop. Do **not block** during your execution!
+    ///
+    /// - Returns: A bool indicating if the request should really be started. Return false if the request has already been cancelled.
+    ///            If the request is cancelled after this method call `executor.cancel()` to stop request execution.
+    func willExecuteRequest(_: HTTPRequestExecutor) -> Bool
+
+    /// Will be called by the ChannelHandler to indicate that the request head has been sent.
+    ///
+    /// This will be called on the Channel's EventLoop. Do **not block** during your execution!
+    func requestHeadSent(_: HTTPRequestHead)
+
+    /// Start request streaming
+    ///
+    /// This will be called on the Channel's EventLoop. Do **not block** during your execution!
+    func startRequestBodyStream()
+
+    /// Pause request streaming
+    ///
+    /// This will be called on the Channel's EventLoop. Do **not block** during your execution!
+    func pauseRequestBodyStream()
+
+    /// Resume request streaming
+    ///
+    /// This will be called on the Channel's EventLoop. Do **not block** during your execution!
+    func resumeRequestBodyStream()
```

Comment:
Why do we need to distinguish between start and resume here? Does it make a difference to the implementor?